### PR TITLE
support set startTime and endTime

### PR DESCRIPTION
**Pull request by beliefgp**
support set startTime and endTime

### CODE REVIEW

**Comment by santigimeno** [comment]
@beliefgp First of all, thanks for the contribution! Before getting into the review, a couple of things I've noticed:
- Would this new options be supported for every kind of scheduling (`RecurrenceRule`, `Object Literal`, `cron-style`)? Or just for `cron-style` scheduling?
- We need tests.

I'll try to review it later today.

**Comment by beliefgp** [comment]
@santigimeno 
1. this new options supported for every kind of scheduling. I have tested it  in my local.
2.I am thinking that how to write the tests. Give me some time。

i am sorry for my english is not good……o(╯□╰)o

**Comment by beliefgp** [comment]
@santigimeno  I have added the tests。

File: lib/schedule.js [hunk 1]
```diff
@@ -186,9 +186,31 @@ Job.prototype.schedule = function(spec) {
...
+
+    if (end != null && !(end instanceof Date) && !isValidDate(end = new Date(end))) {
+      end = null;
+    }
+  }
+
   try {
-    var res = cronParser.parseExpression(spec);
-    inv = scheduleNextRecurrence(res, self);
+    var res = cronParser.parseExpression(spec, { currentDate: start });
```
Comment (santigimeno) [inline lib/schedule.js]: `cronParser` also accepts and `endDate` option. Would it make sense using it here?

File: lib/schedule.js [hunk 2]
```diff
@@ -454,7 +478,7 @@ function prepareNextInvocation() {
       job.stopTrackingInvocation(cinv);
 
       job.invoke();
-      job.emit('run');
+      job.emit('run', inv !== null ? inv.fireDate : null, cinv.endDate);
```
Comment (santigimeno) [inline lib/schedule.js]: why this change?

File: test/start-end-test.js [hunk 3]
```diff
@@ -0,0 +1,313 @@
+'use strict';
+
+var sinon = require('sinon');
+var main = require('../package.json').main;
+var schedule = require('../' + main);
+
+var clock;
+
+module.exports = {
+	setUp: function (cb) {
```
Comment (santigimeno) [inline test/start-end-test.js]: indentation

File: test/start-end-test.js [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+    clock = sinon.useFakeTimers();
+    cb();
+  },
+  'RecurrenceRule': {
+    'no endTime , startTime less than now': function (test) {
+       test.expect(3);
+
+      var job = new schedule.Job(function() {
```
Comment (santigimeno) [inline test/start-end-test.js]: `test.expect(3);` should be aligned with the rest of the test code. The same for every test (I think)

File: test/start-end-test.js [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        rule: rule
+      });
+
+      setTimeout(function() {
+        test.done();
+      }, 3250);
+
+      clock.tick(3250);
+    },
+    'no startTime , endTime less greater now': function (test) {
```
Comment (santigimeno) [inline test/start-end-test.js]: endTime greater than now

File: test/start-end-test.js [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+        rule: '*/1 * * * * *'
+      });
+
+      setTimeout(function() {
+        test.done();
+      }, 3250);
+
+      clock.tick(3250);
+    },
+    'no startTime , endTime less greater now': function (test) {
```
Comment (santigimeno) [inline test/start-end-test.js]: endTime greater than now

File: test/start-end-test.js [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+        rule: { second: null }
+      });
+
+      setTimeout(function() {
+        test.done();
+      }, 3250);
+
+      clock.tick(3250);
+    },
+    'no startTime , endTime less greater now': function (test) {
```
Comment (santigimeno) [inline test/start-end-test.js]: endTime greater than now

File: test/start-end-test.js [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+
+      clock.tick(3250);
+    }
+  },
+  tearDown: function(cb) {
+    clock.restore();
+    cb();
+  }
+};
+
```
Comment (santigimeno) [inline test/start-end-test.js]: remove last line

File: lib/schedule.js (hunk 1 above)
Comment (beliefgp) [inline lib/schedule.js]: when using the endDate cronParser,it will  throw an error as the endDate is less than the startDate.
And execute function scheduleNextRecurrence later, so i think that not using it.
we can handle all the logic of the endDate in function scheduleNextRecurrence.

If you can read chinese, this is a wonderful thing for me，/(ㄒoㄒ)/~~

File: lib/schedule.js (hunk 2 above)
Comment (beliefgp) [inline lib/schedule.js]: I want to know that not run next (arguments is null) or the next run time(arguments is inv.fireDate).

Comment (santigimeno) [inline lib/schedule.js]: I guess it can be useful. A couple of things though: 
- It should go in a different commit as it's a different feature from the startTime/endTime one.
- It should include tests and some kind of documentation ( I know the current documentation is scarce, but we have to start somewhere).

File: lib/schedule.js [hunk 9]
```diff
@@ -186,9 +186,31 @@ Job.prototype.schedule = function(spec) {
...
+    start = spec.start || null;
+    end = spec.end || null;
+    spec = spec.rule;
+
+    if (start != null) {
+      if (!(start instanceof Date)) {
+        start = new Date(start);
+      }
+      if (!isValidDate(start) || start.getTime() < Date.now()) {
+        start = null;
```
Comment (santigimeno) [inline lib/schedule.js]: It can be a good idea adding the `(date instanceof Date)` check inside `isValidDate()` function as they're always used together. It would simplify the code a little.

**Comment by santigimeno** [comment]
It's looking good so far. Thanks.

/cc @albertdatui thoughts on this?

**Comment by santigimeno** [comment]
Pinging @tejasmanohar as well :)

**Comment by albertdatui** [comment]
@santigimeno, It looks good to me. :)

Well done! @beliefgp

**Comment by tejasmanohar** [comment]
lgtm

**Comment by santigimeno** [comment]
@beliefgp if you could address the remaining comments it would be great

**Comment by beliefgp** [comment]
@santigimeno   
1. remove emit,('run') ?  feature add
2. edit isValidDate 

this?

**Comment by santigimeno** [comment]
@beliefgp yes, that was what I had in mind. The 2nd point is not that important and can be included in a following PR if you prefer.

**Comment by beliefgp** [comment]
ok  i delete run first and commit

**Comment by beliefgp** [comment]
@santigimeno   ok, let me think about edit run is good

**Comment by santigimeno** [comment]
LGTM. Can you please squash the PR into a single commit with a descriptive text? Thanks!

**Comment by beliefgp** [comment]
@santigimeno  what mean?  merge all edit, single commit ?

**Comment by albertdatui** [comment]
@beliefgp, yes, merge all of the commits into two commits, the first one is for all the changes and the second one will be for the test cases
